compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1593,7 +1593,7 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
     if (ctx.mode is Mode.Type)
       assignType(cpy.Annotated(tree)(arg1, annot1), arg1, annot1)
     else {
-      val tpt = TypeTree(AnnotatedType(arg1.tpe.widen, Annotation(annot1)))
+      val tpt = TypeTree(AnnotatedType(arg1.tpe, Annotation(annot1)))
```

Comment:
Actually thinking about this more, I think we should keep the `.widen` here.
I think we can only reach this case when the user write:
```scala
val x = 1: @unchecked
```
In this case I think it's fine to interpret this code as if the user had written:
```scala
val x = 1: Int @unchecked
```
If the user really wants to put the annotation on the singleton type, which is rather unusual, he can still explicitly write:
```scala
val x = 1: 1 @unchecked
```
This one will work fine because of the `.widen` which was dropped in `TypeAssigner`

Reply:
But that means #3564 fails, because we have precisely the case where we annotate a singleton with @uncheckedVariance. 

Also, I would argue that someone writing   

    expr: @unchecked

is really annotating the expression `expr`, not its type. The type is an artifact of the way we
express annotated expressions. So it is better that the annotation does _not_ leak into inferred types of the lhs, say. See also #3427, which is another bug that would be fixed by this PR, but would not be fixed by the proposed change.

Reply:
> But that means #3564 fails, because we have precisely the case where we annotate a singleton with @uncheckedvariance.

I don't think so, in #3564 a singleton type is annotated, so `ctx.mode is Type` is true, this case is handled two lines above the current one, and only the `.widen` that was in `TypeAssigner` is relevant there.

> So it is better that the annotation does not leak into inferred types of the lhs, say.

OK. I'm fine with this but note that this is a change of behavior with respect to Scala 2.

Reply:
No, I changed the logic in Typer only, and #3564 started failing again.